In [106]:
# importing required libraries
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from matplotlib.colors import ListedColormap
from sklearn.preprocessing import OrdinalEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier

### Loding the Dataset

In [107]:
# reading the csv file and assigning it to a variable
electrical_grid = pd.read_csv("datasets/Data_for_UCI_named.csv")

In [108]:
# loading the dataset
electrical_grid.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

In [109]:
# getting information about the dataset
electrical_grid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    10000 non-null  float64
 1   tau2    10000 non-null  float64
 2   tau3    10000 non-null  float64
 3   tau4    10000 non-null  float64
 4   p1      10000 non-null  float64
 5   p2      10000 non-null  float64
 6   p3      10000 non-null  float64
 7   p4      10000 non-null  float64
 8   g1      10000 non-null  float64
 9   g2      10000 non-null  float64
 10  g3      10000 non-null  float64
 11  g4      10000 non-null  float64
 12  stab    10000 non-null  float64
 13  stabf   10000 non-null  object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [110]:
# getting a description of the dataset
electrical_grid.describe()

tau1          tau2          tau3          tau4            p1  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean       5.250000      5.250001      5.250004      5.249997      3.750000   
std        2.742548      2.742549      2.742549      2.742556      0.752160   
min        0.500793      0.500141      0.500788      0.500473      1.582590   
25%        2.874892      2.875140      2.875522      2.874950      3.218300   
50%        5.250004      5.249981      5.249979      5.249734      3.751025   
75%        7.624690      7.624893      7.624948      7.624838      4.282420   
max        9.999469      9.999837      9.999450      9.999443      5.864418   

                 p2            p3            p4            g1            g2  \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000   
mean      -1.250000     -1.250000     -1.250000      0.525000      0.525000   
std        0.433035      0.433035      0.433035      0.274256      0.274255   
min       -1.999891     -1.999945     -1.999926      0.050009      0.050053   
25%       -1.624901     -1.625025     -1.624960      0.287521      0.287552   
50%       -1.249966     -1.249974     -1.250007      0.525009      0.525003   
75%       -0.874977     -0.875043     -0.875065      0.762435      0.762490   
max       -0.500108     -0.500072     -0.500025      0.999937      0.999944   

                 g3            g4          stab  
count  10000.000000  10000.000000  10000.000000  
mean       0.525000      0.525000      0.015731  
std        0.274255      0.274255      0.036919  
min        0.050054      0.050028     -0.080760  
25%        0.287514      0.287494     -0.015557  
50%        0.525015      0.525002      0.017142  
75%        0.762440      0.762433      0.044878  
max        0.999982      0.999930      0.109403

In [111]:
# checking for empty fields in the dataset
electrical_grid.isnull().sum()

tau1     0
tau2     0
tau3     0
tau4     0
p1       0
p2       0
p3       0
p4       0
g1       0
g2       0
g3       0
g4       0
stab     0
stabf    0
dtype: int64

since there is no empty field in the dataset, we proceed further

In [112]:
# dropping the "stab" column
new_electrical_grid_df = electrical_grid.drop(["stab"], axis = 1)

In [113]:
# loading the "new_electrical_grid_df" dataframe
new_electrical_grid_df

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
...        ...       ...       ...       ...       ...       ...       ...   
9995  2.930406  9.487627  2.376523  6.187797  3.343416 -0.658054 -1.449106   
9996  3.392299  1.274827  2.954947  6.894759  4.349512 -1.663661 -0.952437   
9997  2.364034  2.842030  8.776391  1.008906  4.299976 -1.380719 -0.943884   
9998  9.631511  3.994398  2.757071  7.821347  2.514755 -0.966330 -0.649915   
9999  6.530527  6.781790  4.349695  8.673138  3.492807 -1.390285 -1.532193   

            p4        g1        g2        g3        g4     stabf  
0    -1.723086  0.650456  0.859578  0.887445  0.958034  unstable  
1    -1.255012  0.413441  0.862414  0.562139  0.781760    stable  
2    -0.920492  0.163041  0.766689  0.839444  0.109853  unstable  
3    -0.997374  0.446209  0.976744  0.929381  0.362718  unstable  
4    -0.554305  0.797110  0.455450  0.656947  0.820923  unstable  
...        ...       ...       ...       ...       ...       ...  
9995 -1.236256  0.601709  0.779642  0.813512  0.608385  unstable  
9996 -1.733414  0.502079  0.567242  0.285880  0.366120    stable  
9997 -1.975373  0.487838  0.986505  0.149286  0.145984    stable  
9998 -0.898510  0.365246  0.587558  0.889118  0.818391  unstable  
9999 -0.570329  0.073056  0.505441  0.378761  0.942631  unstable  

[10000 rows x 13 columns]

In [114]:
# encoding the categorival variable using Ordinal Encoder
ord_enc = OrdinalEncoder()
new_electrical_grid_df["stabf"] = ord_enc.fit_transform(new_electrical_grid_df[["stabf"]])

In [115]:
# loading the encoded dataframe
new_electrical_grid_df

tau1      tau2      tau3      tau4        p1        p2        p3  \
0     2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1     9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2     8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3     0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4     3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   
...        ...       ...       ...       ...       ...       ...       ...   
9995  2.930406  9.487627  2.376523  6.187797  3.343416 -0.658054 -1.449106   
9996  3.392299  1.274827  2.954947  6.894759  4.349512 -1.663661 -0.952437   
9997  2.364034  2.842030  8.776391  1.008906  4.299976 -1.380719 -0.943884   
9998  9.631511  3.994398  2.757071  7.821347  2.514755 -0.966330 -0.649915   
9999  6.530527  6.781790  4.349695  8.673138  3.492807 -1.390285 -1.532193   

            p4        g1        g2        g3        g4  stabf  
0    -1.723086  0.650456  0.859578  0.887445  0.958034    1.0  
1    -1.255012  0.413441  0.862414  0.562139  0.781760    0.0  
2    -0.920492  0.163041  0.766689  0.839444  0.109853    1.0  
3    -0.997374  0.446209  0.976744  0.929381  0.362718    1.0  
4    -0.554305  0.797110  0.455450  0.656947  0.820923    1.0  
...        ...       ...       ...       ...       ...    ...  
9995 -1.236256  0.601709  0.779642  0.813512  0.608385    1.0  
9996 -1.733414  0.502079  0.567242  0.285880  0.366120    0.0  
9997 -1.975373  0.487838  0.986505  0.149286  0.145984    0.0  
9998 -0.898510  0.365246  0.587558  0.889118  0.818391    1.0  
9999 -0.570329  0.073056  0.505441  0.378761  0.942631    1.0  

[10000 rows x 13 columns]

### Declaring the X and Y variables, Splitting and Scaling

In [116]:
# declaring the X and y variables
X = new_electrical_grid_df.iloc[:, :-1].values
y = new_electrical_grid_df.iloc[:, -1].values

# splitting the dataset into the Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)

# carrying out scaling on the dataset before splitting
# applying the Standard Scaler technique to transform the 
# train set (x_train, y_train) and the test set (x_test)
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# fitting Logistic Regression to the Training set
classifier = LogisticRegression(random_state = 1)
classifier.fit(X_train, y_train)

# predicting the Test set results
y_pred = classifier.predict(X_test)

### Evaluating the Logistic Regression Model

#### Confusuin Matrix

In [117]:
# Making the Confusion Matrix    
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[ 494  218]
 [ 174 1114]]


we can see from the confusion matrix results, we have quite alot of correct predictions, in this case which is **1608** (494 + 1114) and **392** (174 + 218) incorrect predictions

### Using scikit learn to train a Random Forest and Extra Trees Classifier

#### Using Random Forest Classifier to train the model

In [118]:
# fitting random forest classification to the Training set
rfc = RandomForestClassifier(random_state = 1)
rfc.fit(X_train, y_train)

# predicting the Test set results
rf_y_pred = classifier.predict(X_test)

##### Evaluating the Logistic Regression Model

In [119]:
# Making the Confusion Matrix  
rfcm = confusion_matrix(y_test, rf_y_pred)
print(rfcm)

[[ 494  218]
 [ 174 1114]]


we can also see that from the confusion matrix results, we have quite alot of correct predictions, in this case which is **1608** (494 + 1114) and **392** (174 + 218) incorrect predictions

#### Using Extra Trees Classifier to train the model

In [120]:
# fitting extra trees classification to the Training set
etc = ExtraTreesClassifier(n_estimators = 50, min_samples_split = 2, min_samples_leaf = 1, max_features = 'auto' )
etc.fit(X_train, y_train)

# predicting the Test set results
et_y_pred = etc.predict(X_test)

#### Evaluating the Logistic Regression Model

In [121]:
# Making the Confusion Matrix  
etcm = confusion_matrix(y_test, et_y_pred)
print(etcm)

[[ 600  112]
 [  43 1245]]


we can see from the confusion matrix of the Extra Tree Classifier, we have more correct predictions than the other two classifiers above. In this case we have **1857** (610 + 1247) correct predictions and **143** (41 + 102) incorrect predictions